### PR TITLE
#167573212 Implement Login Feature

### PR DESCRIPTION
**Pull request by henryade**
#### What does this PR do?
Implement login route

#### Description of Task proposed in this pull request?
- add login route
- test login route
- create user model, migration and seeds file
- add a `findUser` helper function
- add a password helper function
- fix houndCI issue

#### How should this be manually tested (Quality Assurance)?
- check into the branch by running `git checkout ft-login-167573212`
- create a `.env` file, copy the contents of the `.env.sample` file into it and update the fields and also your database details.
- run `npm db:rollback` to migrate and seed into the database.
- run `npm run start:dev` to start the server
- open [postman](https://www.getpostman.com/downloads/) and use seeded to the database to test e.g 
`POST http://localhost:9000/api/v1/sessions/create`
`{`
                `userLogin: 'JhayX'`
                `password: 'incorrect'`
`}`

#### What are the relevant pivotal tracker stories?
- Story ID:  [#167573212](https://www.pivotaltracker.com/story/show/167573212)

#### Any background context you want to add (Operations Impact)?
- run `npm install` to install all dependencies
- update the `.env` file with your database connection string
- run `npm run db:rollback` to create tables in the database and seed 

#### What I have learned working on this feature:
- I learnt how to mock server error test with sinon

#### Screenshots: 
Successful Login:
Screenshot 2019-07-30 at 6 56 22 PM

Unsuccessful login:
Screenshot 2019-07-30 at 6 59 33 PM

### CODE REVIEW

File: .hound.yml
Comment (dinobi) [inline .hound.yml]: Why this change?

File: .sequelizerc
Comment (dinobi) [inline .sequelizerc]: ditto

File: package.json
Comment (dinobi) [inline package.json]: ditto

File: .hound.yml
Comment (henryade) [reply .hound.yml]: To fix the hound probelm.

File: .sequelizerc
Comment (henryade) [reply .sequelizerc]: the seeds script wasn't working. After researching, I found out I had to change the path. I also inform the database code owner.

File: package.json
Comment (henryade) [reply package.json]: the db migrate script was formerly `db:roll` and you asked it to be changed in to `db: migrate` but the change wasn't reflected in the `db:rollback` script before merging. so for the script to work, I had to change it.

File: server/controllers/user.js
Comment (chialuka) [inline server/controllers/user.js]: Please, could you rename this file to upper camel case as it's default export is a class?

Comment (henryade) [reply server/controllers/user.js]: Done

File: server/database/migrations/20190729172658-create-user.js
Comment (dinobi) [inline server/database/migrations/20190729172658-create-user.js]: Why do we have user creation here

File: server/database/models/user.js
Comment (dinobi) [inline server/database/models/user.js]: Why do we have user model here

File: server/database/seeds/20190729181912-user.js
Comment (dinobi) [inline server/database/seeds/20190729181912-user.js]: Why this?

File: server/database/migrations/20190729172658-create-user.js
Comment (henryade) [reply server/database/migrations/20190729172658-create-user.js]: I created it to test the login route with a user seeded into the database

File: server/database/models/user.js
Comment (henryade) [reply server/database/models/user.js]: I created it to test the login route with a user seeded into the database

File: server/database/seeds/20190729181912-user.js
Comment (henryade) [reply server/database/seeds/20190729181912-user.js]: same reason as above which is `I created it to test the login route with a user seeded into the database`

File: test/users/user.test.js
Comment (dinobi) [inline test/users/user.test.js]: rename this to sessions.test.js

File: server/helpers/dateHelper.js
Comment (dinobi) [inline server/helpers/dateHelper.js:14]: Is this result a datetime object?

File: test/users/user.test.js
Comment (henryade) [reply test/users/user.test.js]: okay

File: server/helpers/dateHelper.js
Comment (henryade) [reply server/helpers/dateHelper.js:14]: yes

File: server/database/models/user.js
Comment (dinobi) [reply server/database/models/user.js]: So why did you commit it?

> I committed it so that my teammates and travis can test the route. And also my test coverage will drop if Travis cannot test the new files. Should I will remove it?

File: server/routes/index.js
Comment (dinobi) [inline server/routes/index.js]: what is `user`?

File: server/routes/user.js
Comment (dinobi) [inline server/routes/user.js]: it should be `router.post('/sessions/create', Sessions.create)`
remove the destructing, cos we could possibly have other create methods, like User.create

File: server/routes/index.js
Comment (henryade) [reply server/routes/index.js]: The user router signifying all routes pertaining to user auth. I initially named it auth.js but the team convention is to name the file user.

File: server/routes/user.js
Comment (henryade) [reply server/routes/user.js]: Done

File: server/controllers/Sessions.js
Comment (dinobi) [inline server/controllers/Sessions.js]: What are these?

File: server/controllers/Sessions.js
Comment (dinobi) [inline server/controllers/Sessions.js]: This has been merged now

File: server/controllers/Sessions.js
Comment (dinobi) [inline server/controllers/Sessions.js]: I don't think we need this, just call `toString` on that variable if you must accept a string, regexes are computationally expensive

File: server/controllers/Sessions.js
Comment (dinobi) [inline server/controllers/Sessions.js:37]: How about we start thinking of models as objects that can hold properties, which is what they are. It would be nice to have this method that query a User table on the associated model object. I need all `rambo-devs` to reason in this light.

File: server/controllers/Sessions.js
Comment (dinobi) [inline server/controllers/Sessions.js]: please move this logic to a helper

File: server/controllers/Sessions.js
Comment (dinobi) [inline server/controllers/Sessions.js]: update your PR

File: server/controllers/Sessions.js
Comment (dinobi) [inline server/controllers/Sessions.js]: Is the token present in this response?, and also added to response header

**Review by dinobi — CHANGES_REQUESTED** [review]
Please update your PR and test

**Comment by henryade** [comment]
> Please update your PR and test

**Review by chialuka — DISMISSED** [review]

**Review by ayodejiAA — DISMISSED** [review]
Looks okay. Good work

File: server/docs/authors-haven-api.yml
Comment (abdulfatai360) [inline server/docs/authors-haven-api.yml:74]: Ade, I am aware that there were no validation rules at the time you wrote this documentation. However, there is an adopted rule now. Could you please reflect them here in your documentation?

File: server/controllers/Sessions.js
Comment (abdulfatai360) [inline server/controllers/Sessions.js:38]: @henryade Could you please check if the request body is empty or either of the user login details is not present and return the appropriate response to the client?

**Review by chialuka — APPROVED** [review]

**Review by Adekoreday — APPROVED** [review]

**Review by JuwonAbiola — APPROVED** [review]

**Review by abdulfatai360 — APPROVED** [review]

**Pull request merged**
